### PR TITLE
Send Notification/Reminder messages to ActionExporter and Notify-gateway

### PR DESCRIPTION
**Pull request by jwgwalton**
- Populates the classifiers and statuses required for sending notifications/generating print letters.
- Validates these messages against the business rules for whether they are to be sent to handler services for letters and emails.

### CODE REVIEW

**Comment by benjefferies** [comment]
@jwgwalton might be good to add a link to the API changes in the description so they can be merged first

**Comment by jwgwalton** [comment]
Disagree with the rest of the codacy issues.

File: src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java [hunk 1]
```diff
@@ -252,6 +248,21 @@ private ActionRequest createActionRequest(final Action action, final ActionPlan
     actionRequest.setSurveyName(surveyDTO.getLongName());
     actionRequest.setSurveyRef(surveyDTO.getSurveyRef());
 
+    CommsTypeClassifiers classifiers = new CommsTypeClassifiers();
+    classifiers.setLegalBasis(surveyDTO.getLegalBasis());
+    classifiers.setRegion(businessUnitAttributes.getRegion());
```
Comment (jwgwalton) [inline src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java]: @pricem14pc I was thinking about the blank region code and wasn't sure if we wanted an internal representation to map these to?

File: src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java [hunk 2]
```diff
@@ -261,6 +272,52 @@ private ActionRequest createActionRequest(final Action action, final ActionPlan
...
+   * @return enrolment status
+   */
+  private String getEnrolmentStatus(final PartyDTO childParty) {
+    List<String> enrolmentStatuses = new ArrayList<>();
+    for (Association association : childParty.getAssociations()) {
+      for (Enrolment enrolment : association.getEnrolments()) {
+        enrolmentStatuses.add(enrolment.getEnrolmentStatus());
+      }
+    }
+    if (enrolmentStatuses.contains(ENABLED)) {
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java]: What determines the priority of enabled vs pending etc. If the enrolmentStatuses contains all three why is enabled chosen first? Maybe a test around this would be handy

Comment (jwgwalton) [reply src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java]: You could have multiple respondents which is why there is an order of precedence. I will add a test thought.

File: src/test/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImplTest.java [hunk 3]
```diff
@@ -489,4 +507,118 @@ public void testProcessActionCancelHappyPath() throws CTPException {
     assertTrue(publishedActionCancel.isResponseRequired());
     assertEquals(CANCELLATION_REASON, publishedActionCancel.getReason());
   }
+
+  @Test
+  public void testActionInstructionNotSentIfInvalid() throws CTPException{
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImplTest.java]: Do we need to have all of the when steps defined? Do we need all of the verify steps defined; do we care what functions are called other than that `sendActionInstruction` is not sent? I ask because it's a bit hard to follow the tests. If we don't it could be cut down to something like
```java
  @Test
  public void testActionInstructionNotSentIfInvalid() throws CTPException{
    // Start of section to mock responses
    when(caseSvcClientService.getCaseWithIACandCaseEvents(CASE_ID)).thenReturn(caseDetailsDTOs.get(0));
    when(partySvcClientService.getParty(SAMPLE_UNIT_TYPE_H, PARTY_ID)).thenReturn(partyDTOs.get(0));
    when(collectionExerciseClientService.getCollectionExercise(COLLECTION_EXERCISE_ID)). thenReturn(collectionExerciseDTOs.get(0));
    when(surveySvcClientService.requestDetailsForSurvey(CENSUS)).thenReturn(surveyDTOs.get(0));

    // End of section to mock responses

    // Start of section to run the test
    Action action = Action.builder()
            .id(ACTION_ID)
            .actionType(ActionType.builder().responseRequired(Boolean.TRUE).handler(ACTIONEXPORTER).build())
            .caseId(CASE_ID)
            .priority(1).build();
    actionProcessingService.processActionRequest(action);
    // End of section to run the test

    // VALIDATOR HAS STOPPED MESSAGE FROM BEING SENT
    verify(actionInstructionPublisher, times(0)).sendActionInstruction(eq(ACTIONEXPORTER),
            any(uk.gov.ons.ctp.response.action.message.instruction.Action.class));
  }
```

File: src/test/java/uk/gov/ons/ctp/response/action/service/impl/ActionRequestValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+import static org.junit.Assert.assertFalse;
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+import uk.gov.ons.ctp.response.action.domain.model.ActionType;
+import uk.gov.ons.ctp.response.action.message.instruction.ActionRequest;
+import uk.gov.ons.ctp.response.casesvc.representation.CaseGroupStatus;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class ActionRequestValidatorTest {
```
Comment (benjefferies) [inline src/test/java/uk/gov/ons/ctp/response/action/service/impl/ActionRequestValidatorTest.java]: Nice readable tests 👍

**Comment by benjefferies** [comment]
@jwgwalton How can I test this locally? Is it ok to add some testing instructions to the PR description?

File: src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java [hunk 5]
```diff
@@ -261,6 +277,95 @@ private ActionRequest createActionRequest(final Action action, final ActionPlan
...
+    return enrolmentStatus;
+  }
+
+  /**
+   * Concatenate the businessUnitAttributes trading style fields into a single string
+   * @param businessUnitAttributes
+   * @return concatenated trading styles
+   */
+  private String generateTradingStyle(final Attributes businessUnitAttributes) {
+    String tradStyle1 = businessUnitAttributes.getTradstyle1();
```
[the remark this answers is not in the document]
Comment (jwgwalton) [reply src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java]: Put it in and put in some specific tests for it to make up for any loss in legibility

File: src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java [hunk 6]
```diff
@@ -261,6 +277,95 @@ private ActionRequest createActionRequest(final Action action, final ActionPlan
...
+
+    List<String> childPartyIds = parentParty.getAssociations().stream().map(Association::getPartyId).collect(Collectors.toList());
+
+    List<String> childPartyStatuses = new ArrayList<>();
+
+    //TODO: THIS IS AWFUL!! All child samples are parent+I but must be better way
+    String childUnitTypeStr = parentUnitTypeStr + "I";
+
+    for (String id : childPartyIds) {
+      try {
```
[the remark this answers is not in the document]
Comment (jwgwalton) [reply src/main/java/uk/gov/ons/ctp/response/action/service/impl/ActionProcessingServiceImpl.java]: Doesn't throw an exception, I'm just a mug.

File: pom.xml [hunk 7]
```diff
@@ -31,19 +31,19 @@
     <dependency>
       <groupId>uk.gov.ons.ctp.product</groupId>
       <artifactId>partysvc-api</artifactId>
-      <version>10.50.2</version>
+      <version>10.50.3</version>
```
Comment (rjweeks) [inline pom.xml]: Need to change this to 10.50.4

Comment (jwgwalton) [reply pom.xml]: the release version with the update to the json ignore is 10.50.3, travis has bumped the snapshot version up to to 10.50.4 for development.
